### PR TITLE
Save to WebAnnotator format using html walk instead of xml sax

### PR DESCRIPTION
**Pull request by whalebot-helmsman**
Using html walk we can preserve html comments and do not annotate script/style tags

### CODE REVIEW

File: webstruct/tests/test_webannotator.py [hunk 1]
```diff
@@ -91,15 +256,15 @@ def test_wa_convert(self):
             <p>
               __START_ORG__ Scrapinghub
                 <b>Inc __END_ORG__  has</b>an
-                <b>__START_CITY__office</b>in Montevideo __END_CITY__   __START_ORG__ SH __END_ORG__
+                <b> __START_CITY__ office</b>in Montevideo __END_CITY__   __START_ORG__ SH __END_ORG__ cool
```
Comment (kmike) [inline webstruct/tests/test_webannotator.py]: Why is space added before start token?

File: webstruct/webannotator.py [hunk 2]
```diff
@@ -264,6 +186,107 @@ def _set_base(tree, baseurl):
...
+    if is_tail:
+        source = element.tail
+
+    if not source or not source.strip():
+        return
+
+    remainder = source[:first.position]
+
+    nodes = list()
+    for number, (start, end, _id) in enumerate(tasks):
```
Comment (kmike) [inline webstruct/webannotator.py]: number is a very uncommon name for index variables; would you mind using "idx" or "index"?

File: webstruct/tests/test_webannotator.py (hunk 1 above)
Comment (whalebot-helmsman) [reply webstruct/tests/test_webannotator.py]: Just for strictness. In html tokenizer we use ' __START/END_SMTH__ '(with spaces) tags, so in towebannotator function we should look for tags with spaces.

File: webstruct/webannotator.py [hunk 3]
```diff
@@ -264,6 +186,107 @@ def _set_base(tree, baseurl):
     head = _ensure_head(tree)
     head.insert(0, Element("base", href=baseurl))
 
+TagPosition = namedtuple('TagPosition', ['element',
+                                         'tag',
+                                         'position',
+                                         'length',
+                                         'is_tail',
+                                         'dfs_number'])
```
Comment (kmike) [inline webstruct/webannotator.py]: I think "dfs_number" implies something like "amount of 'dfs'" or "count of 'dfs'", but likely this is not what is meant here - is it an index/position of some kind?

File: webstruct/webannotator.py [hunk 4]
```diff
@@ -264,6 +186,107 @@ def _set_base(tree, baseurl):
     head = _ensure_head(tree)
     head.insert(0, Element("base", href=baseurl))
 
+TagPosition = namedtuple('TagPosition', ['element',
```
Comment (kmike) [inline webstruct/webannotator.py]: It seems this namedtuple is an implementation detail of to_webannotator function - what do you think about making it private, i.e. renaming to `_TagPosition`?

File: webstruct/webannotator.py [hunk 5]
```diff
@@ -264,6 +186,107 @@ def _set_base(tree, baseurl):
...
+def fabricate_start(element, is_tail, tag):
+    return TagPosition(element=element,
+                       tag=tag,
+                       position=0,
+                       length=0,
+                       is_tail=is_tail,
+                       dfs_number=0)
+
+
+def fabricate_end(element, is_tail, tag):
```
Comment (kmike) [inline webstruct/webannotator.py]: fabricate_* functions look pretty internal, I think they should be private

**Comment by kmike** [comment]
Thanks @whalebot-helmsman! I like the changes in functionality, but for me it is quite hard to follow the code. Could you please help me by renaming some variables, extracting functions, maybe adding a few high-level comments about the overall approach?

File: webstruct/webannotator.py [hunk 6]
```diff
@@ -280,11 +303,113 @@ def to_webannotator(tree, entity_colors=None, url=None):
...
+    ends = [e for e in translate_to_dfs(ends, ordered)]
+
+    starts.sort(key=lambda t: (t.dfs_number, t.position))
+    ends.sort(key=lambda t: (t.dfs_number, t.position))
+
+    dfs_order = (max(n for n in ordered.values()) + 1) * [None]
+    for text_node, dfs_number in ordered.items():
+        dfs_order[dfs_number] = text_node
+
+    tasks = []
```
Comment (kmike) [inline webstruct/webannotator.py]: Can we rename this variable, to make code more clear? Is it a list of positions to enclose with WebAnnotator spans? Something like positions_to_enclose would be more readable. There was even a variable with the same name within the same function, but with a different meaning.

File: webstruct/webannotator.py [hunk 7]
```diff
@@ -280,11 +303,113 @@ def to_webannotator(tree, entity_colors=None, url=None):
...
+        for text_node in dfs_order[start_number + 1:end_number]:
+            if text_node is None:
+                continue
+
+            element, is_tail = text_node
+            is_string = isinstance(element.tag, six.string_types)
+            is_text = isinstance(element.tag, six.text_type)
+
+            if not is_string and not is_text:
+                continue
```
Comment (kmike) [inline webstruct/webannotator.py]: six.text_type check is redundant, as text_type is unicode, and string_types includes both bytes and unicode in Python2, or just unicode in Python 3. It seems you can just check `if isinstance(element.tag, six.string_types)`.

File: webstruct/webannotator.py [hunk 8]
```diff
@@ -280,11 +303,113 @@ def to_webannotator(tree, entity_colors=None, url=None):
...
+        enclose(enclosures, entity_colors)
+
+    _copy_title(root)
+    _add_wacolor_elements(root, entity_colors)
     if url is not None:
-        _set_base(tree, url)
-    return tree
+        _set_base(root, url)
+
+    return root
```
Comment (kmike) [inline webstruct/webannotator.py]: Could you extracting separate functions from `to_webannotator`? For example, `_copy_title` and `_add_wacolor_elements` calls are definitely not at the same level of abstraction as two etree.iterwalk loops.

**Comment by whalebot-helmsman** [comment]
Just an explanation of what is happening in this code. The hardest question to answer in this transformation is "If we have start/stop tags for same entity in different text/tail nodes which text/tail node we should annotate too?"
I propose next answer. We should walk the DOM tree in depth first order and number all nodes. Also we number each text node as first child and tail node as last child. So when we have start tag in node with number _n_ and stop tag in node with number _n+m_, we should annotate all nodes with numbers _n+i_ , where _i_ in range _[1,m)_

File: webstruct/webannotator.py (hunk 3 above)
Comment (whalebot-helmsman) [reply webstruct/webannotator.py]: When we talk about depth first search it is common to call these indices numbers. [Wiki article](https://en.wikipedia.org/wiki/Depth-first_search#DFS_ordering) references whole process as enumeration.

Comment (kmike) [reply webstruct/webannotator.py]: @whalebot-helmsman ok! Well, if you search "number" in this article, it won't have the meaning you're using here - it is only used as "number of nodes", "number of edges". But I can find "dfs number" mentioned in other articles, so that's fair to keep this name, you're right.

File: webstruct/webannotator.py (hunk 7 above)
Comment (whalebot-helmsman) [reply webstruct/webannotator.py]: Actually it has no sense as there is no Unicode tags at all. lxml even does not handle them properly. So string type is enough.

Comment (kmike) [reply webstruct/webannotator.py]: six.string_type includes unicode anyways

File: webstruct/webannotator.py [hunk 9]
```diff
@@ -280,11 +388,47 @@ def to_webannotator(tree, entity_colors=None, url=None):
...
+        raise ValueError('len(ends) != len(starts)')
+
+    ordered = _enumerate_nodes_in_dfs_order(root)
+    starts = [s for s in _translate_to_dfs(starts, ordered)]
+    ends = [e for e in _translate_to_dfs(ends, ordered)]
+
+    starts.sort(key=lambda t: (t.dfs_number, t.position))
+    ends.sort(key=lambda t: (t.dfs_number, t.position))
+
+    dfs_order = (max(n for n in ordered.values()) + 1) * [None]
```
Comment (kmike) [inline webstruct/webannotator.py]: why not just `max(ordered.values())`?

**Comment by kmike** [comment]
The algorithm is much more clear now, thanks!

File: webstruct/webannotator.py (hunk 9 above)
Comment (whalebot-helmsman) [reply webstruct/webannotator.py]: It is not first variant. Such form is a relic of previous versions.
